Source/JavaScriptCore/lol/LOLJIT.h

```diff
@@ -0,0 +1,335 @@
+    macro(op_nstricteq) \
+    macro(op_get_prototype_of) \
+    macro(op_check_tdz) \
+    macro(op_to_property_key) \
+    macro(op_to_property_key_or_number) \
+    macro(op_typeof_is_function) \
+
+
+// We subclass ReplayBackend so we can just pass `this` to the ReplayAllocator as we can't use `{ }` for the allocator's internal Backend& in the constructor.
+class LOLJIT : public JIT, public ReplayBackend {
+    WTF_MAKE_TZONE_NON_HEAP_ALLOCATABLE(LOLJIT);
+public:
+
+    LOLJIT(VM&, BaselineJITPlan&, CodeBlock*);
+
+    RefPtr<BaselineJITCode> compileAndLinkWithoutFinalizing(JITCompilationEffort);
+
+private:
+    template<typename> friend class RegisterAllocator;
+    void fill(VirtualRegister binding, GPRReg gpr)
+    {
+        JIT_COMMENT(*this, "Filling ", binding);
+        emitGetVirtualRegister(binding, gpr);
+    }
+
+    void flush(const Location& location, GPRReg gpr, VirtualRegister binding)
+    {
+        JIT_COMMENT(*this, "Flushing ", binding);
+        if (!location.isFlushed)
+            emitPutVirtualRegister(binding, gpr);
+#if ASSERT_ENABLED
+        else {
+            JIT_COMMENT(*this, " already flushed, validating");
+            emitGetVirtualRegister(binding, scratchRegister());
+            Jump ok = branch64(Equal, scratchRegister(), gpr);
+            breakpoint();
+            ok.link(this);
+        }
+#endif
+    }
+
+    ALWAYS_INLINE constexpr static bool hasSlowCase(OpcodeID op)
+    {
+        switch (op) {
+#define HAS_SLOW_CASE(name) case name: return true;
+        FOR_EACH_OP_WITH_OPERATION_SLOW_CASE(HAS_SLOW_CASE)
+        FOR_EACH_OP_WITH_SLOW_CASE(HAS_SLOW_CASE)
+        default: break;
+#undef HAS_SLOW_CASE
+        }
+        return false;
+    }
+
+    ALWAYS_INLINE constexpr static bool isImplemented(OpcodeID op)
+    {
+        switch (op) {
+#define HAS_SLOW_CASE(name) case name: return true;
+        FOR_EACH_IMPLEMENTED_OP(HAS_SLOW_CASE)
+        default: break;
+#undef HAS_SLOW_CASE
+        }
+        return false;
+    }
+
+private:
+    void privateCompileMainPass();
+    void privateCompileSlowCases();
+
+#define DECLARE_EMIT_METHODS(name) \
+    void emit_##name(const JSInstruction*); \
+    void emitSlow_##name(const JSInstruction*, Vector<SlowCaseEntry>::iterator&);
+    FOR_EACH_IMPLEMENTED_OP(DECLARE_EMIT_METHODS)
+#undef DECLARE_EMIT_METHODS
+
+    void nextBytecodeIndexWithFlushForJumpTargetsIfNeeded(auto& allocator, bool shouldSetFastPathResumePoint)
+    {
+        auto next = BytecodeIndex(m_bytecodeIndex.offset() + m_currentInstruction->size());
+        if (m_currentJumpTargetIndex < m_unlinkedCodeBlock->numberOfJumpTargets() && next.offset() == m_unlinkedCodeBlock->jumpTarget(m_currentJumpTargetIndex)) {
+            if (shouldSetFastPathResumePoint) {
+                // We need to set a resume point for slow paths to jump back to prior to flushing since the next instruction wouldn't have the flushes and we don't want to re-emit them in the slow path.
```

Comment:
I think that it's worth having a comment somewhere that illustrates what the emitted code is intended to look like / what the control flow of a slow path that needs to flush would be